src/script/interpreter.cpp

```diff
@@ -1963,8 +1989,10 @@ static bool ExecuteWitnessScript(const Span<const valtype>& stack_span, const CS
                 // Note how this condition would not be reached if an unknown OP_SUCCESSx was found
                 return set_error(serror, SCRIPT_ERR_BAD_OPCODE);
             }
+
+            const bool op_cat_is_op_success = !(flags & SCRIPT_VERIFY_TAPSCRIPT_OP_CAT);
             // New opcodes will be listed here. May use a different sigversion to modify existing opcodes.
-            if (IsOpSuccess(opcode)) {
+            if (IsOpSuccess(opcode) || (op_cat_is_op_success && opcode == OP_CAT)) {
```

Comment:
Sorry, got this logic wrong. The idea is that each "SCRIPT_VERIFY" flag should be a soft-fork -- so adding an extra flag should only make previously valid scripts invalid, and never make a previously invalid script valid. This fails that convention since `DISCOURAGE_OP_SUCCESS` would consider `01 02 CAT` to be an invalid script, but `DISCOURAGE_OP_SUCCESS | TAPSCRIPT_OP_CAT` would consider it to be valid.

Should be:

```c++
    if (IsOpSuccess(opcode)) {
        if (opcode == OP_CAT) {
            if (flags & DISCOURAGE_TAPSCRIPT_OP_CAT) {
                return set_error(...);
            } else if (!(flags & TAPSCRIPT_OP_CAT)) {
                // op_success behaviour
                return set_success(serror);
            }
        } else {
            // old OP_SUCCESS code
            if (flags & DISCOURAGE_OP_SUCCESS) {
                return set_error(serror, SCRIPT_ERR_DISCOURAGE_OP_SUCCESS);
            }
            return set_success(serror);
        }
    }
```

(leave `IsOpSuccess` untouched, and put the `opcode == OP_CAT` test inside the if, that way you're structurally guaranteed this is a soft fork)

Reply:
Ah I think I understand. Thanks for catching that.

Reply:
Added this fix to the last commit & force pushed